### PR TITLE
kv: replace memdb with a more memory efficient version

### PR DESCRIPTION
**Pull request by bobotu**
### What problem does this PR solve? 
The existing memdb from goleveldb use a contiguous memory to store data. It will put pressure on system memory when enlarging the underlying slice. After supporting large transaction, this problem will be more obvious.

### What is changed and how it works?
Implementing a new skiplist with a block-based memory allocator, when enlarging the memdb simply allocate a new block instead of allocating a large amount of memory and copy all of the existing data into the new memory location.

The block size starts at an initial block size then doubled when each new block allocated until it reaches 128M. 

Because there are some internal data structure optimization, the performance of `Get` and `Put` is slightly improved.

Here is the result of benchmarks in `kv` package:
### master
```
BenchmarkIsPoint-16                     300000000                4.84 ns/op            0 B/op          0 allocs/op
BenchmarkMemDbBufferSequential-16             20          61710345 ns/op        11979968 B/op          3 allocs/op
BenchmarkMemDbBufferRandom-16                 10         116388720 ns/op        13374355 B/op          7 allocs/op
BenchmarkMemDbIter-16                        300           5030566 ns/op             208 B/op          3 allocs/op
BenchmarkMemDbCreation-16                3000000               426 ns/op            4096 B/op          1 allocs/op
```
### new
```
BenchmarkIsPoint-16                     300000000                4.81 ns/op            0 B/op          0 allocs/op
BenchmarkMemDbBufferSequential-16             30          60371850 ns/op         8947658 B/op          1 allocs/op
BenchmarkMemDbBufferRandom-16                 20         101129510 ns/op         6710522 B/op          1 allocs/op
BenchmarkMemDbIter-16                       1000           1220734 ns/op             128 B/op          1 allocs/op      
BenchmarkMemDbCreation-16                5000000               394 ns/op            4096 B/op          1 allocs/op  
```

### master
```
BenchmarkIsPoint-16                     300000000                4.82 ns/op            0 B/op          0 allocs/op
BenchmarkMemDbBufferSequential-16              1        7888385900 ns/op        3561011328 B/op       96 allocs/op
BenchmarkMemDbBufferRandom-16                  1        46474700000 ns/op       3561046784 B/op       98 allocs/op
BenchmarkMemDbIter-16                          2         504151600 ns/op             208 B/op          3 allocs/op
BenchmarkMemDbCreation-16                3000000               417 ns/op            4096 B/op          1 allocs/op
```
### new
```
BenchmarkIsPoint-16                     300000000                4.82 ns/op            0 B/op          0 allocs/op
BenchmarkMemDbBufferSequential-16              1        7038679100 ns/op        536864032 B/op        39 allocs/op
BenchmarkMemDbBufferRandom-16                  1        36407376900 ns/op       536864032 B/op        39 allocs/op
BenchmarkMemDbIter-16                         10         122093460 ns/op             128 B/op          1 allocs/op
BenchmarkMemDbCreation-16                5000000               400 ns/op            4096 B/op          1 allocs/op
```
### Check List 

Tests 

 - Unit test

Side effects

 - Increased code complexity

### CODE REVIEW

File: kv/memdb/memdb.go [hunk 1]
```diff
@@ -0,0 +1,398 @@
...
+			// before.key < key < next.key. We are done for this level.
+			return before, next, cmp == 0
+		}
+		before = next // Keep moving right on this level.
+	}
+}
+
+func (db *DB) findGreater(key []byte, allowEqual bool) (*node, []byte, bool) {
+	var prev *node
+	prev = db.head.node
```
Comment (ngaut) [inline kv/memdb/memdb.go]: use assignment pattern.

File: kv/memdb/memdb.go [hunk 2]
```diff
@@ -0,0 +1,398 @@
...
+			continue
+		}
+		return next, nextData, false
+	}
+}
+
+func (db *DB) findLess(key []byte, allowEqual bool) (*node, []byte, bool) {
+	var prev *node
+	var prevData []byte
+	prev = db.head.node
```
Comment (ngaut) [inline kv/memdb/memdb.go]: Ditto.

File: kv/memdb/memdb.go [hunk 3]
```diff
@@ -0,0 +1,398 @@
...
+	}
+	return prev, prevData, false
+}
+
+// findLast returns the last element. If head (empty db), we return nil. All the find functions
+// will NEVER return the head nodes.
+func (db *DB) findLast() (*node, []byte) {
+	var node *node
+	var nodeData []byte
+	node = db.head.node
```
Comment (ngaut) [inline kv/memdb/memdb.go]: Ditto.

File: kv/memdb/memdb.go [hunk 4]
```diff
@@ -0,0 +1,395 @@
...
+	if next[0].node != nil {
+		next[0].prev = uint64(addr)
+	}
+
+	db.length++
+	db.size += len(key) + len(v)
+	return true
+}
+
+func (db *DB) prepareOverwrite(next []nodeWithAddr) int {
```
Comment (coocood) [inline kv/memdb/memdb.go]: Need comment for this function.

File: kv/memdb/memdb.go [hunk 5]
```diff
@@ -0,0 +1,395 @@
...
+	old := next[0]
+	height := int(old.height)
+	db.size -= int(old.valLen) + int(old.keyLen)
```
Comment (coocood) [inline kv/memdb/memdb.go]: need update `db.length`?

File: kv/memdb/arena.go [hunk 6]
```diff
@@ -0,0 +1,119 @@
...
+func (addr arenaAddr) blockIdx() int {
+	return int(addr>>32 - 1)
+}
+
+func (addr arenaAddr) blockOffset() uint32 {
+	return uint32(addr)
+}
+
+func newArenaAddr(blockIdx int, blockOffset uint32) arenaAddr {
+	return arenaAddr(uint64(blockIdx+1)<<32 | uint64(blockOffset))
```
[the remark this answers is not in the document]
Comment (bobotu) [reply kv/memdb/arena.go]: `blockIdx == 0 && blockOffset == 0` means `null`.
Maybe we can place a `nil` pointer at `blocks[0]`  to make `blockIdx` starts at `1`?

**Comment by coocood** [comment]
LGTM

File: kv/memdb/memdb.go [hunk 7]
```diff
@@ -200,7 +200,7 @@ func (db *DB) findSpliceForLevel(arena *arena, key []byte, before nodeWithAddr,
 	}
 }
 
-func (db *DB) findGreater(key []byte, allowEqual bool) (*node, []byte, bool) {
+func (db *DB) findGreater(key []byte) (*node, []byte, bool) {
```
Comment (coocood) [inline kv/memdb/memdb.go]: findGreaterEqual?

File: kv/memdb/memdb_test.go [hunk 8]
```diff
@@ -170,8 +175,130 @@ func (s testMemDBSuite) TestDelete(c *C) {
...
+	c.Check(it.Value(), BytesEquals, key)
+	it.Prev()
+	c.Check(it.Valid(), IsFalse)
+}
+
+func (s testMemDBSuite) TestRandom(c *C) {
+	const cnt = 500000
+	p1 := New(4 * 1024)
+	p2 := memdb.New(comparer.DefaultComparer, 4*1024)
+	var buf [4]byte
```
Comment (coocood) [inline kv/memdb/memdb_test.go]: Can we make the key length random?

File: kv/memdb/arena.go [hunk 9]
```diff
@@ -0,0 +1,117 @@
...
+func (a *arena) alloc(size int) (arenaAddr, []byte) {
+	if size > a.blockSize {
+		// Use a separate block to store entry which size larger than specified block size.
+		blk := newArenaBlock(size)
+		addr := newArenaAddr(len(a.blocks), 0)
+		a.blocks = append(a.blocks, blk)
+		return addr, blk.buf
+	}
+
+	for {
```
Comment (tiancaiamao) [inline kv/memdb/arena.go]: if size > maxBlockSize, this is a dead loop and lead to OOM eventually?

File: kv/memdb/arena.go [hunk 10]
```diff
@@ -0,0 +1,117 @@
...
+		}
+		a.blocks = append(a.blocks, newArenaBlock(a.blockSize))
+		a.availIdx = int(uint32(len(a.blocks) - 1))
+	}
+}
+
+func (a *arena) reset() {
+	a.availIdx = 0
+	a.blockSize = len(a.blocks[0].buf)
+	a.blocks = []*arenaBlock{a.blocks[0]}
```
Comment (tiancaiamao) [inline kv/memdb/arena.go]: The reset drop all `arenaBlock`? So what's the benefit compared to create a new arena?

File: kv/memdb/memdb.go [hunk 11]
```diff
@@ -0,0 +1,364 @@
...
+func New(initBlockSize int) *DB {
+	return &DB{
+		height: 1,
+		head:   nodeWithAddr{node: new(node)},
+		arena:  newArenaLocator(initBlockSize),
+	}
+}
+
+// Reset resets the DB to initial empty state.
+// Release all blocks except the init one.
```
Comment (tiancaiamao) [inline kv/memdb/memdb.go]: So Reset has no memory re-use...

File: kv/memdb/memdb.s [hunk 12]
```diff
@@ -0,0 +1,12 @@
+// Copyright 2019 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
```
Comment (tiancaiamao) [inline kv/memdb/memdb.s]: No content in this file?

Comment (bobotu) [reply kv/memdb/memdb.s]: Empty file to enable `go:linkname fastRand runtime.fastrand` at `memdb.go:L310`

File: kv/memdb/arena.go (hunk 9 above)
Comment (bobotu) [reply kv/memdb/arena.go]: `a.blockSize <= maxBlockSize` and `alloc` checked `if size > a.blockSize` at L57. So `size > maxBlockSize` never step into this loop.

**Comment by tiancaiamao** [comment]
LGTM

**Pull request merged**